`restore-packages.cmd` for Python on Windows broken due to Numpy 2 release, breaks Boost build

## What is the issue
Build pipeline breaks due to failure to compile boost on Windows

Building Boost (b2.exe <options) fails with following console logging.
<details closed>
<summary>Boost Console Output with `--verbose` flag</summary>
<br>

> 
> [1] msvc-14.2
> 
> Component configuration:
> 
>     - atomic                   : not building
>     - charconv                 : not building
>     - chrono                   : not building
>     - cobalt                   : not building
>     - container                : not building
>     - context                  : not building
>     - contract                 : not building
>     - coroutine                : not building
>     - date_time                : not building
>     - exception                : not building
>     - fiber                    : not building
>     - filesystem               : not building
>     - graph                    : not building
>     - graph_parallel           : not building
>     - headers                  : not building
>     - iostreams                : not building
>     - json                     : not building
>     - locale                   : not building
>     - log                      : not building
>     - math                     : not building
>     - mpi                      : not building
>     - nowide                   : not building
>     - program_options          : not building
>     - python                   : building
>     - random                   : not building
>     - regex                    : not building
>     - serialization            : not building
>     - stacktrace               : not building
>     - system                   : not building
>     - test                     : not building
>     - thread                   : not building
>     - timer                    : not building
>     - type_erasure             : not building
>     - url                      : not building
>     - wave                     : not building
> 
> ...patience...
> ...patience...
> ...found 35849 targets...
> ...updating 5 targets...
> compile-c-c++ bin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi\numpy\dtype.obj
> dtype.cpp
> libs\python\src\numpy\dtype.cpp(101): error C2039: 'elsize': is not a member of '_PyArray_Descr'
> C:\Python310\lib\site-packages\numpy\_core\include\numpy\ndarraytypes.h(618): note: see declaration of '_PyArray_Descr'
> 
>     call "bin.v2\standalone\msvc\msvc-14.2\msvc-setup.bat"  >nul
>  cl /Zm800 -nologo "libs\python\src\numpy\dtype.cpp" -c -Fo"bin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi\numpy\dtype.obj"     -TP /wd4675 /EHs /GR /Zc:throwingNew /Z7 /Od /Ob0 /W3 /MDd /Zc:forScope /Zc:wchar_t /Zc:inline /favor:blend -DBOOST_ALL_NO_LIB=1 -DBOOST_COBALT_USE_STD_PMR=1 -DBOOST_NUMPY_SOURCE -DBOOST_NUMPY_STATIC_LIB -DBOOST_PYTHON_STATIC_LIB "-I." "-IC:\Python310\include" "-IC:\Python310\lib\site-packages\numpy\_core\include"
> 
> ...failed compile-c-c++ bin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi\numpy\dtype.obj...
> ...skipped <pbin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi>libboost_numpy310-vc142-mt-gd-x64-1_85.lib for lack of <pbin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi>numpy\dtype.obj...
> ...skipped <pC:\Users\seleonar.NORTHAMERICA\Documents\Dev\lang_ext\packages\output\lib>libboost_numpy310-vc142-mt-gd-x64-1_85.lib for lack of <pbin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi>libboost_numpy310-vc142-mt-gd-x64-1_85.lib...
> ...skipped <pbin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi>libboost_numpy-variant-vc142-mt-gd-x64-1_85-static-py3.10.cmake for lack of <pbin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi>libboost_numpy310-vc142-mt-gd-x64-1_85.lib...
> ...skipped <pC:\Users\seleonar.NORTHAMERICA\Documents\Dev\lang_ext\packages\output\lib\cmake\boost_numpy-1.85.0>libboost_numpy-variant-vc142-mt-gd-x64-1_85-static-py3.10.cmake for lack of <pbin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi>libboost_numpy-variant-vc142-mt-gd-x64-1_85-static-py3.10.cmake...
> 
> ...skipped 4 targets...
>    <pC:\Users\seleonar.NORTHAMERICA\Documents\Dev\lang_ext\packages\output\lib>libboost_numpy310-vc142-mt-gd-x64-1_85.lib
>    <pC:\Users\seleonar.NORTHAMERICA\Documents\Dev\lang_ext\packages\output\lib\cmake\boost_numpy-1.85.0>libboost_numpy-variant-vc142-mt-gd-x64-1_85-static-py3.10.cmake
>    <pbin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi>libboost_numpy-variant-vc142-mt-gd-x64-1_85-static-py3.10.cmake
>    <pbin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi>libboost_numpy310-vc142-mt-gd-x64-1_85.lib
> 
> ...failed updating 1 target...
>    compile-c-c++ bin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi\numpy\dtype.obj

</details>

The following error message led me to the root cause:
> compile-c-c++ bin.v2\libs\python\build\msvc-14.2\debug\link-static\python-3.10\threading-multi\numpy\dtype.obj
dtype.cpp
libs\python\src\numpy\dtype.cpp(101): error C2039: 'elsize': is not a member of '_PyArray_Descr'
C:\Python310\lib\site-packages\numpy_core\include\numpy\ndarraytypes.h(618): note: see declaration of '_PyArray_Descr'

Which then led me to finding that the following two files differ where Windows does not explicitly set Numpy and Pandas versions while the Linux script does: (Numpy-1.22.3 and Panda-1.4.2)
- "C:\enlistment\language-extensions\python\build\linux\restore-packages.sh"
- "C:\enlistment\language-extensions\python\build\windows\restore-packages.cmd"

### External Reporting 

- https://github.com/boostorg/python/issues/431
- https://github.com/boostorg/python/issues/442
- https://github.com/boostorg/python/issues/441

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -61,7 +61,7 @@ REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive
 REM -o Output directory
 REM 2nd param is the file to expand
 echo -- Beginning Boost ZIP extraction -- Time: %time% --
-%ARCHIVE_TOOL_PATH% e -y -o"%PACKAGES_ROOT%" "boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
+%ARCHIVE_TOOL_PATH% e -y -o"%PACKAGES_ROOT%/boost_%BOOST_VERSION_IN_UNDERSCORE%" "boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
 echo -- Finished Boost Zip extration -- Time: %time% --
 
 REM Boost cleanup

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -88,7 +88,6 @@ echo -- Beginning Boost b2.exe build -- Time: %time% --
 CALL bootstrap.bat vc142
 echo -- Beginning Boost build using compiled b2.exe-- Time: %time% --
 b2.exe -j12 --prefix=%PACKAGES_ROOT%\output --with-python --user-config="%PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%\user-config.jam" --debug-configuration -dp0 toolset=msvc-14.2
-REM  address-model=64 variant=debug link=static threading=multi runtime-link=shared install
 echo -- Finished Boost build -- Time: %time% --
 
 REM If building in pipeline, set the PYTHONHOME here to overwrite the existing PYTHONHOME

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -4,7 +4,7 @@ CALL %ENL_ROOT%\restore-packages.cmd
 SET PACKAGES_ROOT=%ENL_ROOT%\packages
 
 REM 7zip file extraction tooling - Direct path on pipeline
-SET 7ZIPPATH="C:\Program Files\7-Zip\7z.exe"
+SET 7ZIPPATH="C:\7-Zip\7z.exe"
 
 REM Specify the Python version to be downloaded and installed
 REM

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -43,7 +43,9 @@ curl -sS https://bootstrap.pypa.io/get-pip.py |"%PYTHON_INSTALLATION_PATH%\pytho
 
 REM Install numpy and pandas
 REM
-"%PYTHON_INSTALLATION_PATH%\python.exe" -m pip install pandas numpy
+SET NUMPY_VERSION=1.22.3
+SET PANDAS_VERSION=1.4.2
+"%PYTHON_INSTALLATION_PATH%\python.exe" -m pip install --force-reinstall numpy==%NUMPY_VERSION% pandas==%PANDAS_VERSION%
 
 REM Remove the Python installer which is no longer needed
 REM
@@ -56,7 +58,6 @@ REM Remove the Boost zip file
 REM Navigate to the extracted Boost directory
 REM
 curl -L -o boost_%BOOST_VERSION_IN_UNDERSCORE%.7z https://archives.boost.io/release/%BOOST_VERSION%/source/boost_%BOOST_VERSION_IN_UNDERSCORE%.7z
-REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive -Force -Path 'boost_%BOOST_VERSION_IN_UNDERSCORE%.zip' -DestinationPath '%PACKAGES_ROOT%'"
 
 REM -o Output directory
 REM 2nd param is the file to expand
@@ -80,11 +81,14 @@ REM
 echo using python : %PYTHON_VERSION_MAJOR_MINOR% : "%PYTHON_INSTALLATION_PATH_DOUBLE_SLASH%\\python" : "%PYTHON_INSTALLATION_PATH_DOUBLE_SLASH%\\include" : "%PYTHON_INSTALLATION_PATH_DOUBLE_SLASH%\\libs" ; > user-config.jam
 
 REM Run Boost's bootstrap script and build Boost.Python with the created configuration
-REM
+REM Explicitly set VS2019 Compiler tooling 
+REM bootstrap - "vc142"
+REM b2.exe "toolset=msvc-14.2"
 echo -- Beginning Boost b2.exe build -- Time: %time% --
-CALL bootstrap.bat
+CALL bootstrap.bat vc142
 echo -- Beginning Boost build using compiled b2.exe-- Time: %time% --
-b2.exe -j12 --with-python --user-config="%PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%\user-config.jam" --debug-configuration -d0
+b2.exe -j12 --prefix=%PACKAGES_ROOT%\output --with-python --user-config="%PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%\user-config.jam" --debug-configuration -dp0 toolset=msvc-14.2
+REM  address-model=64 variant=debug link=static threading=multi runtime-link=shared install
 echo -- Finished Boost build -- Time: %time% --
 
 REM If building in pipeline, set the PYTHONHOME here to overwrite the existing PYTHONHOME

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -4,7 +4,7 @@ CALL %ENL_ROOT%\restore-packages.cmd
 SET PACKAGES_ROOT=%ENL_ROOT%\packages
 
 REM 7zip file extraction tooling - Direct path on pipeline
-SET ARCHIVE_TOOL_PATH="C:\Program Files\7-Zip\7z.exe"
+SET ARCHIVE_TOOL_PATH="C:\7-Zip\7z.exe"
 
 REM Specify the Python version to be downloaded and installed
 REM

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -61,7 +61,7 @@ REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive
 REM -o Output directory
 REM 2nd param is the file to expand
 echo -- Beginning Boost ZIP extraction -- Time: %time% --
-%ARCHIVE_TOOL_PATH% e -y -o"%PACKAGES_ROOT%" "%PACKAGES_ROOT%/boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
+%ARCHIVE_TOOL_PATH% e -y -o"%PACKAGES_ROOT%" "boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
 echo -- Finished Boost Zip extration -- Time: %time% --
 
 REM Boost cleanup

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -4,7 +4,7 @@ CALL %ENL_ROOT%\restore-packages.cmd
 SET PACKAGES_ROOT=%ENL_ROOT%\packages
 
 REM 7zip file extraction tooling - Direct path on pipeline
-SET 7ZIPPATH="C:\7-Zip\7z.exe"
+SET ARCHIVE_TOOL_PATH=C:\7-Zip\7z.exe
 
 REM Specify the Python version to be downloaded and installed
 REM
@@ -61,7 +61,7 @@ REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive
 REM -o Output directory
 REM 2nd param is the file to expand
 echo -- Beginning Boost ZIP extraction -- Time: %time% --
-%7ZIPPATH% e -y -o"%PACKAGES_ROOT%" "%PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
+%ARCHIVE_TOOL_PATH% e -y -o"%PACKAGES_ROOT%" "%PACKAGES_ROOT%/boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
 echo -- Finished Boost Zip extration -- Time: %time% --
 
 REM Boost cleanup

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -4,7 +4,7 @@ CALL %ENL_ROOT%\restore-packages.cmd
 SET PACKAGES_ROOT=%ENL_ROOT%\packages
 
 REM 7zip file extraction tooling - Direct path on pipeline
-SET ARCHIVE_TOOL_PATH=C:\7-Zip\7z.exe
+SET ARCHIVE_TOOL_PATH="C:\Program Files\7-Zip\7z.exe"
 
 REM Specify the Python version to be downloaded and installed
 REM
@@ -61,7 +61,7 @@ REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive
 REM -o Output directory
 REM 2nd param is the file to expand
 echo -- Beginning Boost ZIP extraction -- Time: %time% --
-%ARCHIVE_TOOL_PATH% e -y -o"%PACKAGES_ROOT%/boost_%BOOST_VERSION_IN_UNDERSCORE%" "boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
+%ARCHIVE_TOOL_PATH% x -y -o"packages" "boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
 echo -- Finished Boost Zip extration -- Time: %time% --
 
 REM Boost cleanup
@@ -74,8 +74,9 @@ echo using python : %PYTHON_VERSION_MAJOR_MINOR% : "%PYTHON_INSTALLATION_PATH_DO
 
 REM Run Boost's bootstrap script and build Boost.Python with the created configuration
 REM
-echo -- Beginning Boost build using b2.exe-- Time: %time% --
+echo -- Beginning Boost b2.exe build -- Time: %time% --
 CALL bootstrap.bat
+echo -- Beginning Boost build using compiled b2.exe-- Time: %time% --
 b2.exe -j12 --with-python --user-config="%PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%\user-config.jam" --debug-configuration -d0
 echo -- Finished Boost build -- Time: %time% --
 

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -65,9 +65,15 @@ echo -- Beginning Boost ZIP extraction -- Time: %time% --
 echo -- Finished Boost Zip extration -- Time: %time% --
 
 REM Boost cleanup
+echo Delete 7z boost archive
 del boost_%BOOST_VERSION_IN_UNDERSCORE%.7z
+echo go to dir with boost unarchived
+%cd%
+dir
 pushd %PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%
-
+%cd%
+dir
+echo create user config jam
 REM Create a Boost user-config.jam configuration file for building Boost.Python
 REM
 echo using python : %PYTHON_VERSION_MAJOR_MINOR% : "%PYTHON_INSTALLATION_PATH_DOUBLE_SLASH%\\python" : "%PYTHON_INSTALLATION_PATH_DOUBLE_SLASH%\\include" : "%PYTHON_INSTALLATION_PATH_DOUBLE_SLASH%\\libs" ; > user-config.jam

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -52,7 +52,12 @@ REM Remove the Boost zip file
 REM Navigate to the extracted Boost directory
 REM
 curl -L -o boost_%BOOST_VERSION_IN_UNDERSCORE%.zip https://sourceforge.net/projects/boost/files/boost/%BOOST_VERSION%/boost_%BOOST_VERSION_IN_UNDERSCORE%.zip/download
-powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive -Force -Path 'boost_%BOOST_VERSION_IN_UNDERSCORE%.zip' -DestinationPath '%PACKAGES_ROOT%'"
+REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive -Force -Path 'boost_%BOOST_VERSION_IN_UNDERSCORE%.zip' -DestinationPath '%PACKAGES_ROOT%'"
+
+REM -o Output directory
+REM 2nd param is the file to expand
+C:\7-Zip\7z.exe e -y -o"%PACKAGES_ROOT%" "boost_%BOOST_VERSION_IN_UNDERSCORE%.zip"
+
 del boost_%BOOST_VERSION_IN_UNDERSCORE%.zip
 pushd %PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%
 

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -61,17 +61,18 @@ REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive
 REM -o Output directory
 REM 2nd param is the file to expand
 echo -- Beginning Boost ZIP extraction -- Time: %time% --
-%ARCHIVE_TOOL_PATH% x -y -o"packages" "boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
+%ARCHIVE_TOOL_PATH% x -y -o"%PACKAGES_ROOT%" "boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
 echo -- Finished Boost Zip extration -- Time: %time% --
 
 REM Boost cleanup
 echo Delete 7z boost archive
 del boost_%BOOST_VERSION_IN_UNDERSCORE%.7z
 echo go to dir with boost unarchived
-%cd%
+echo %cd%
 dir
+echo %PACKAGES_ROOT%
 pushd %PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%
-%cd%
+echo %cd%
 dir
 echo create user config jam
 REM Create a Boost user-config.jam configuration file for building Boost.Python

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -3,6 +3,9 @@ CALL %ENL_ROOT%\restore-packages.cmd
 
 SET PACKAGES_ROOT=%ENL_ROOT%\packages
 
+REM 7zip file extraction tooling - Direct path on pipeline
+SET 7ZIPPATH="C:\Program Files\7-Zip\7z.exe"
+
 REM Specify the Python version to be downloaded and installed
 REM
 SET PYTHON_VERSION=3.10.2
@@ -46,20 +49,23 @@ REM Remove the Python installer which is no longer needed
 REM
 del "python-%PYTHON_VERSION%.exe"
 
+REM BOOST artifact download, extract, build
 REM Download the specified version of Boost from SourceForge
 REM Extract the downloaded Boost zip file to the packages directory
 REM Remove the Boost zip file
 REM Navigate to the extracted Boost directory
 REM
-curl -L -o boost_%BOOST_VERSION_IN_UNDERSCORE%.zip https://sourceforge.net/projects/boost/files/boost/%BOOST_VERSION%/boost_%BOOST_VERSION_IN_UNDERSCORE%.zip/download
+curl -L -o boost_%BOOST_VERSION_IN_UNDERSCORE%.7z https://archives.boost.io/release/%BOOST_VERSION%/source/boost_%BOOST_VERSION_IN_UNDERSCORE%.7z
 REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive -Force -Path 'boost_%BOOST_VERSION_IN_UNDERSCORE%.zip' -DestinationPath '%PACKAGES_ROOT%'"
 
 REM -o Output directory
 REM 2nd param is the file to expand
 echo -- Beginning Boost ZIP extraction -- Time: %time% --
-C:\7-Zip\7z.exe e -y -o"%PACKAGES_ROOT%" "boost_%BOOST_VERSION_IN_UNDERSCORE%.zip"
+%7ZIPPATH% e -y -o"%PACKAGES_ROOT%" "%PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%.7z"
 echo -- Finished Boost Zip extration -- Time: %time% --
-del boost_%BOOST_VERSION_IN_UNDERSCORE%.zip
+
+REM Boost cleanup
+del boost_%BOOST_VERSION_IN_UNDERSCORE%.7z
 pushd %PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%
 
 REM Create a Boost user-config.jam configuration file for building Boost.Python
@@ -68,8 +74,10 @@ echo using python : %PYTHON_VERSION_MAJOR_MINOR% : "%PYTHON_INSTALLATION_PATH_DO
 
 REM Run Boost's bootstrap script and build Boost.Python with the created configuration
 REM
+echo -- Beginning Boost build using b2.exe-- Time: %time% --
 CALL bootstrap.bat
 b2.exe -j12 --with-python --user-config="%PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%\user-config.jam" --debug-configuration -d0
+echo -- Finished Boost build -- Time: %time% --
 
 REM If building in pipeline, set the PYTHONHOME here to overwrite the existing PYTHONHOME
 REM

--- a/language-extensions/python/build/windows/restore-packages.cmd
+++ b/language-extensions/python/build/windows/restore-packages.cmd
@@ -56,8 +56,9 @@ REM powershell -NoProfile -ExecutionPolicy Unrestricted -Command "Expand-Archive
 
 REM -o Output directory
 REM 2nd param is the file to expand
+echo -- Beginning Boost ZIP extraction -- Time: %time% --
 C:\7-Zip\7z.exe e -y -o"%PACKAGES_ROOT%" "boost_%BOOST_VERSION_IN_UNDERSCORE%.zip"
-
+echo -- Finished Boost Zip extration -- Time: %time% --
 del boost_%BOOST_VERSION_IN_UNDERSCORE%.zip
 pushd %PACKAGES_ROOT%\boost_%BOOST_VERSION_IN_UNDERSCORE%
 

